## Parameters for all Patients 

In [1]:
from __future__ import division, print_function

%matplotlib inline
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import pandas as pd
from collections import OrderedDict
from scipy.signal import savgol_filter
from scipy.interpolate import InterpolatedUnivariateSpline

Massage spreadsheet data into a dictionary of patients indexed by integer participant number and with each entry a dataframe with LA and LV volumes indexed as time from trigger in millisecs.

There is also a dataframe of RR and frame intervals intervals indexed on participant number.

In [2]:
volumes = 'FullVolumes.xlsx'
rrintervals = 'RRIntervals.xlsx'

df_vols = pd.read_excel(volumes)
df_rrintervals = pd.read_excel(rrintervals).reset_index(drop=True)
df_rrintervals.index = ['RR_ms']

df_la = df_vols.filter(regex='LA-[0-9]{3,3}$')
df_lv = df_vols.filter(regex='LV-[0-9]{3,3}$')

# map index to integers
df_times = df_rrintervals.filter(regex='LV-[0-9]{3,3}$').T
df_times['PN'] = [int(i[-3:]) for i in df_times.index]
df_times = df_times.set_index('PN')

nrows = df_la.count(axis=0)[0]
df_times['Frame_interval_ms'] = df_times['RR_ms'] / nrows
patients = {}
for patno in [int(col[-3:]) for col in df_la]:
    patients[patno] = pd.concat([
        pd.Series(np.arange(nrows) * df_times['Frame_interval_ms'][patno], name='Time'),
        df_lv['LV-%d' % patno],
        df_la['LA-%d' % patno],
    ], axis=1).set_index('Time')
    patients[patno].columns = ('LV', 'LA')
    patients[patno].pat_name = 'Patient %d' % patno

Parameters of Interest. Work on numpy arrays of volumes without the frame rate so all time values are indices ie in 'frames' rather than millisecs.

In [3]:
def lv_es_index(volumes):
    return np.argmin(volumes)

def lv_split_at_es(volumes):
    es = lv_es_index(volumes)
    return volumes[:es], volumes[es:]

def lv_vol_at_es(volumes):
    return volumes[lv_es_index(volumes)]

def lv_vol_at_ss(volumes):
    return volumes[0]

def lv_vol_at_ed(volumes):
    return volumes[-1]

def idx_nearest(array, value, after=0):
    return np.abs(array[after:]-value).argmin() + after

def lv_idx_for_refill80_full(volumes, use_vss=True):
    vss = lv_vol_at_ss(volumes)
    ves = lv_vol_at_es(volumes)
    ved = lv_vol_at_ed(volumes)
    
    fullvol = vss if use_vss else (vss + ved) / 2
    emptyvol = ves
    
    eighty_vol = 0.2*emptyvol + 0.8*fullvol
    return idx_nearest(volumes, value=eighty_vol, after=lv_es_index(volumes))

def lv_idx_max_systolic_down_slope(volumes, slopes):
    esi = lv_es_index(volumes)
    systolic_down_slopes = -slopes[:esi]
    return systolic_down_slopes.argmax()

def lv_idx_max_recovery_slopes(volumes, slopes):
    esi = lv_es_index(volumes)
    # need to further split into early and late.
    idx80 = lv_idx_for_refill80_full(volumes)
    
    diastolic_early_up_slopes = slopes[esi:idx80]
    diastolic_late_up_slopes = slopes[idx80:]
    idx_early = diastolic_early_up_slopes.argmax() + esi
    idx_late = diastolic_late_up_slopes.argmax() + idx80
    
    return idx_early, idx_late

def la_es_index(volumes):
    return np.argmax(volumes)

def la_split_at_es(volumes):
    es = la_es_index(volumes)
    return volumes[:es], volumes[es:]

def la_vol_at_es(volumes):
    return volumes[la_es_index(volumes)]

def la_vol_at_es(volumes, es):
    return volumes[es]

def la_vol_at_ss(volumes):
    return volumes[0]

def la_vol_at_ed(volumes):
    return volumes[-1]

def la_idx_max_systolic_up_slope(volumes, slopes):
    esi = la_es_index(volumes)
    systolic_down_slopes = slopes[:esi]
    return systolic_down_slopes.argmax()

def la_idx_max_emptying_slopes(volumes, slopes):
    esi = la_es_index(volumes)

    # need to further split into early and late.
    idxsplit = esi + (len(volumes) - esi) // 2
    
    diastolic_early_up_slopes = -slopes[esi:idxsplit]
    diastolic_late_up_slopes = -slopes[idxsplit:]
    idx_early = diastolic_early_up_slopes.argmax() + esi
    idx_late = diastolic_late_up_slopes.argmax() + idxsplit
    
    return idx_early, idx_late


In [4]:
def lv_params(patient, region='LV'):
    units, time_units = 'ml', 'ms'
    patient_name = patient.pat_name
    
    volume_series = patient[region]
    time = np.asarray(volume_series.index)
    volume = np.asarray(volume_series.values) 
    diffs = np.gradient(volume) / np.diff(time)[0]
    r_to_r = time[-1]
    sg_volume     = savgol_filter(x=volume, window_length=5, polyorder=2, mode='interp')
    dt_msecs      = np.mean(np.diff(time))
    # one order higher for the derivative sounds right .. increase window to match
    sg_derivative = savgol_filter(x=volume, window_length=7, polyorder=3, deriv=1, delta=dt_msecs)

    spline = InterpolatedUnivariateSpline(time, sg_volume)
    spline_deriv = InterpolatedUnivariateSpline(time, sg_derivative)

    interpolated_time   = np.linspace(0, time[-1], len(time)*3)
    interpolated_savgol = spline(interpolated_time)
    interpolated_deriv  = spline_deriv(interpolated_time)

    min_vol_index = lv_es_index(interpolated_savgol)
    min_vol_time  = interpolated_time[min_vol_index]
    min_volume    = interpolated_savgol[min_vol_index]
    eighty_percent_idx = lv_idx_for_refill80_full(interpolated_savgol)
    eighty_percent_time = interpolated_time[eighty_percent_idx]
    eighty_percent_vol = interpolated_savgol[eighty_percent_idx]

    max_emptying_index = lv_idx_max_systolic_down_slope(interpolated_savgol, interpolated_deriv)
    max_emptying_time = interpolated_time[max_emptying_index]
    max_emptying_vol = interpolated_savgol[max_emptying_index]
    max_emptying_slope = interpolated_deriv[max_emptying_index]

    max_early_filling_index, max_late_filling_index = lv_idx_max_recovery_slopes(interpolated_savgol, interpolated_deriv)
    max_early_filling_time = interpolated_time[max_early_filling_index]
    max_late_filling_time = interpolated_time[max_late_filling_index]

    max_early_filling_vol = interpolated_savgol[max_early_filling_index]
    max_late_filling_vol = interpolated_savgol[max_late_filling_index]

    max_early_filling_slope = interpolated_deriv[max_early_filling_index]    
    max_late_filling_slope = interpolated_deriv[max_late_filling_index]    
    diastolic_time = r_to_r - min_vol_time
    
    return OrderedDict([
        ('LV_time_end_systole', min_vol_time),
        ('LV_peak_ejection_ml_sec', (-1000 *max_emptying_slope)),
        ('LV_time_peak_ejection',  max_emptying_time),
        ('LV_vol_peak_ejection', max_emptying_vol),
        ('LV_early_peak_filling_ml_sec', (1000 *max_early_filling_slope)),
        ('LV_time_early_peak_filling', (max_early_filling_time - min_vol_time)),
        ('LV_vol_early_peak_filling', max_early_filling_vol),
        ('LV_late_peak_filling_ml_sec', (1000 *max_late_filling_slope)),
        ('LV_time_late_peak_filling', (max_late_filling_time - min_vol_time)),
        ('LV_vol_late_peak_filling', max_late_filling_vol),
        ('LV_time_fill80', (eighty_percent_time - min_vol_time)),
        ('LV_ratio_fill80', (eighty_percent_time - min_vol_time) / diastolic_time)
   ])


In [5]:
params178 = lv_params(patients[178])

In [6]:
def la_params(patient, region='LA'):
    units, time_units = 'ml', 'ms'
    patient_name = patient.pat_name
    
    volume_series = patient[region]
    time = np.asarray(volume_series.index)
    volume = np.asarray(volume_series.values) 
    diffs = np.gradient(volume) / np.diff(time)[0]
    r_to_r = time[-1]
    sg_volume     = savgol_filter(x=volume, window_length=5, polyorder=2, mode='interp')
    dt_msecs      = np.mean(np.diff(time))
    # one order higher for the derivative sounds right .. increase window to match
    sg_derivative = savgol_filter(x=volume, window_length=7, polyorder=3, deriv=1, mode='interp', delta=dt_msecs)

    spline = InterpolatedUnivariateSpline(time, sg_volume)
    spline_deriv = InterpolatedUnivariateSpline(time, sg_derivative)

    interpolated_time   = np.linspace(0, time[-1], len(time)*3)
    interpolated_savgol = spline(interpolated_time)
    interpolated_deriv  = spline_deriv(interpolated_time)

    max_vol_index = la_es_index(interpolated_savgol)
    max_vol_time  = interpolated_time[max_vol_index]
    max_volume    = interpolated_savgol[max_vol_index]

    max_filling_index = la_idx_max_systolic_up_slope(interpolated_savgol, interpolated_deriv)
    max_filling_time = interpolated_time[max_filling_index]
    max_filling_vol = interpolated_savgol[max_filling_index]
    max_filling_slope = interpolated_deriv[max_filling_index]
    
    max_early_emptying_index, max_late_emptying_index = la_idx_max_emptying_slopes(interpolated_savgol, interpolated_deriv)

    max_early_emptying_time = interpolated_time[max_early_emptying_index]
    max_early_emptying_vol = interpolated_savgol[max_early_emptying_index]
    max_early_emptying_slope = interpolated_deriv[max_early_emptying_index]    

    max_late_emptying_time = interpolated_time[max_late_emptying_index]
    max_late_emptying_vol = interpolated_savgol[max_late_emptying_index]
    max_late_emptying_slope = interpolated_deriv[max_late_emptying_index]    

    return OrderedDict([
        ('LA_time_end_systole', max_vol_time),
        ('LA_peak_filling_ml_sec', (1000 * max_filling_slope)),
        ('LA_time_peak_filling', max_filling_time),
        ('LA_vol_peak_filling', max_filling_vol),
        ('LA_early_peak_emptying_ml_sec', (-1000 *max_early_emptying_slope)),
        ('LA_time_early_peak_emptying', (max_early_emptying_time - max_vol_time)),
        ('LA_vol_early_peak_emptying', max_early_emptying_vol),
        ('LA_late_peak_emptying_ml_sec', (-1000 *max_late_emptying_slope)),
        ('LA_time_late_peak_emptying', (max_late_emptying_time - max_vol_time)),
        ('LA_vol_late_peak_emptying', max_late_emptying_vol),
    ])

In [7]:
params178.update(la_params(patients[178]))

In [8]:
def all_params(patient):
    params =lv_params(patient)
    params.update(la_params(patient))
    return params

results = {k: all_params(p) for (k, p) in patients.items()}

In [9]:
df_results = pd.DataFrame(results).T
df_results.insert(0, 'RR_ms', df_times['RR_ms'])
df_results.head(20).T

178         179         182         183  \
RR_ms                          674.157303  845.070423  952.380952  821.917808   
LA_early_peak_emptying_ml_sec  276.998490  224.700038  163.888526  104.980410   
LA_late_peak_emptying_ml_sec   158.269147  396.176009  612.500000  343.394511   
LA_peak_filling_ml_sec         178.906512  270.301828  226.458333  172.075452   
LA_time_early_peak_emptying     52.474947   54.815379   49.420849  106.627175   
LA_time_end_systole            288.612208  405.633803  790.733591  373.195113   
LA_time_late_peak_emptying     288.612208  405.633803  123.552124  415.845983   
LA_time_peak_filling            78.712420  219.261515    0.000000  106.627175   
LA_vol_early_peak_emptying      47.864520   94.689106   99.811614   56.778191   
LA_vol_late_peak_emptying       31.811313   45.250857   73.400000   43.885714   
LA_vol_peak_filling             37.776525   80.143452   66.485714   44.217654   
LV_early_peak_filling_ml_sec   523.506899  482.415697  444.518490  339.377459   
LV_late_peak_filling_ml_sec    236.911109  656.544560  549.062500  502.719907   
LV_peak_ejection_ml_sec        798.853638  754.916797  551.558899  448.831896   
LV_ratio_fill80                  0.652174    0.937500    1.000000    0.980392   
LV_time_early_peak_filling     113.695718  175.409212  605.405405  522.473158   
LV_time_end_systole            244.883085  285.039970  296.525097  245.242503   
LV_time_fill80                 262.374734  493.338409  617.760618  533.135876   
LV_time_late_peak_filling      306.103857  526.227636  617.760618  543.798593   
LV_time_peak_ejection           87.458245  109.630758  135.907336  117.289893   
LV_vol_early_peak_filling       88.669452   81.846384  145.995928   78.972851   
LV_vol_late_peak_filling       150.180118  180.718000  150.742857   86.200000   
LV_vol_peak_ejection           122.287323  146.851530  127.027676   62.699107   

                                      186          187         188  \
RR_ms                          983.606557  1090.909091  714.285714   
LA_early_peak_emptying_ml_sec  363.414202   207.494257  278.300678   
LA_late_peak_emptying_ml_sec   434.806548   189.131014  183.611111   
LA_peak_filling_ml_sec         267.977527   131.652611  216.039106   
LA_time_early_peak_emptying     89.322109    70.761671  101.930502   
LA_time_end_systole            433.850244   424.570025  342.857143   
LA_time_late_peak_emptying     510.412051   594.398034  342.857143   
LA_time_peak_filling           255.206026   353.808354   74.131274   
LA_vol_early_peak_emptying     114.483429    61.028417   67.159955   
LA_vol_late_peak_emptying       61.742857    33.396440   37.257143   
LA_vol_peak_filling            108.959769    64.389000   49.041482   
LV_early_peak_filling_ml_sec   547.698858   393.818082  549.874258   
LV_late_peak_filling_ml_sec    303.890542   182.611536  385.416667   
LV_peak_ejection_ml_sec        538.335206   380.801676  566.084338   
LV_ratio_fill80                  0.875000     0.826087    0.837838   
LV_time_early_peak_filling     204.164821   113.218673  120.463320   
LV_time_end_systole            331.767833   396.265356  342.857143   
LV_time_fill80                 535.932654   537.788698  287.258687   
LV_time_late_peak_filling      612.494462   551.941032  342.857143   
LV_time_peak_ejection          216.925122   127.371007  148.262548   
LV_vol_early_peak_filling      104.850052    66.150736   86.583128   
LV_vol_late_peak_filling       172.914286   114.900197  146.685714   
LV_vol_peak_ejection           101.908281    98.836987  101.985308   

                                      191         192          198  \
RR_ms                          845.070423  895.522388  1034.482759   
LA_early_peak_emptying_ml_sec  228.763436   43.300268    80.217240   
LA_late_peak_emptying_ml_sec   142.929646  553.785797   284.150132   
LA_peak_filling_ml_sec         153.127985  405.071941   159.112633   
LA_time_early_peak_emptying     65.778455   92.940702   228.14

We need some quality control - some values are dubious here: eg 182, 192, 209 ...

 - 182
   - LV
     - fails to recover to 80%
     - misses early filling as as later is steeper
   - LA
     - missidentifies ES as second peak higher
     - final slope down too steep as seems just at final point (edge behaviour of SG?)

- Need something better to distinguish early and late phases ...
  - could get multiple peaks with some sort of peak finder and choose two highest
  - will need some sort of robust peak finder ....

In [10]:
df_results[df_results['LV_ratio_fill80'] > 0.95]

RR_ms  LA_early_peak_emptying_ml_sec  LA_late_peak_emptying_ml_sec  \
182   952.380952                     163.888526                    612.500000   
183   821.917808                     104.980410                    343.394511   
240  1153.846154                      51.845238                    142.123016   
266  1276.595745                     244.678488                    478.315146   
300  1016.949153                     106.065748                    291.097884   

     LA_peak_filling_ml_sec  LA_time_early_peak_emptying  LA_time_end_systole  \
182              226.458333                    49.420849           790.733591   
183              172.075452                   106.627175           373.195113   
240              138.531624                    59.875260           493.970894   
266              229.754390                   115.928695           447.153537   
300              153.195571                   158.314246           448.557032   

     LA_time_late_peak_emptying  LA_time_peak_filling  \
182                  123.552124              0.000000   
183                  415.845983            106.627175   
240                  613.721414            149.688150   
266                  778.378378            298.102358   
300                  527.714155            145.121393   

     LA_vol_early_peak_emptying  LA_vol_late_peak_emptying  ...  \
182                   99.811614                  73.400000  ...   
183                   56.778191                  43.885714  ...   
240                   87.714286                  67.942857  ...   
266                  108.156828                  68.685714  ...   
300                  117.817181                 100.742857  ...   

     LV_peak_ejection_ml_sec  LV_ratio_fill80  LV_time_early_peak_filling  \
182               551.558899         1.000000                  605.405405   
183               448.831896         0.980392                  522.473158   
240               542.258402         1.000000                  808.316008   
266               496.382070         0.980392                  182.173663   
300               493.975274         0.958333                  211.085662   

     LV_time_end_systole  LV_time_fill80  LV_time_late_peak_filling  \
182           296.525097      617.760618                 617.760618   
183           245.242503      533.135876                 543.798593   
240           284.407484      823.284823                 823.284823   
266           380.908568      828.062105                 844.623347   
300           343.014201      606.871278                 633.256986   

     LV_time_peak_ejection  LV_vol_early_peak_filling  \
182             135.907336                 145.995928   
183             117.289893                  78.972851   
240             164.656965                  87.011940   
266             149.051179                  74.425549   
300             158.314246                  79.659052   

     LV_vol_late_peak_filling  LV_vol_peak_ejection  
182                150.742857            127.027676  
183                 86.200000             62.699107  
240                 91.085714             67.085295  
266                145.371429            109.273580  
300                115.028571            101.227491  

[5 rows x 23 columns]

----